<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>FineST in NPC Visium data &mdash; FineST 0.0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css" />
      <link rel="stylesheet" type="text/css" href="_static/nbsphinx-code-cells.css" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="FineST in CRC Visium HD data" href="CRC16_Train_Impute.html" />
    <link rel="prev" title="Crop image of ROI" href="Crop_ROI_image.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            FineST
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Main</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="StarDist_demo.html">Nuclei segmentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="HIPT_demo.html">Geometric segmentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="release.html">Release History</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Between_spot_demo.html">Interpolate between-spots</a></li>
<li class="toctree-l1"><a class="reference internal" href="Crop_ROI_image.html">Crop image of ROI</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">FineST in NPC Visium data</a></li>
<li class="toctree-l1"><a class="reference internal" href="CRC16_Train_Impute.html">FineST in CRC Visium HD data</a></li>
<li class="toctree-l1"><a class="reference internal" href="NPC_LRI_CCC.html">LRI and CCC discovery in super-resolved NPC data</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">FineST</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">FineST in NPC Visium data</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/NPC_Train_Impute.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="FineST-in-NPC-Visium-data">
<h1>FineST in NPC Visium data<a class="headerlink" href="#FineST-in-NPC-Visium-data" title="Permalink to this heading"></a></h1>
<section id="Main-steps">
<h2>Main steps<a class="headerlink" href="#Main-steps" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>Step1: Prepare paired image patch and ST spot: Align image pixels with spot coordinates</p></li>
<li><p>Step2: Train model on within spots: Train the model on 80% data and test on 20% data</p></li>
<li><p>Step3: Infer on between spots: Infer gene expression of between spot using trained model</p></li>
<li><p>Step4: Impute on all sub-spots: Impute gene expression using measured gene expression</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">torch</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">datetime</span> <span class="kn">import</span> <span class="n">datetime</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">matplotlib.colors</span> <span class="k">as</span> <span class="nn">clr</span>
<span class="kn">import</span> <span class="nn">json</span>

<span class="kn">import</span> <span class="nn">FineST</span> <span class="k">as</span> <span class="nn">fst</span>
<span class="kn">from</span> <span class="nn">FineST.datasets</span> <span class="kn">import</span> <span class="n">dataset</span>
<span class="kn">import</span> <span class="nn">FineST.plottings</span> <span class="k">as</span> <span class="nn">fstplt</span>
<span class="nb">print</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">__version__</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;FineST version: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span><span class="k">fst</span>.__version__)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
1.7.1
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## From GPU2</span>
<span class="c1"># path = &#39;/mnt/lingyu/nfs_share2/Python/&#39;</span>
<span class="c1"># os.chdir(str(path) + &#39;FineST/FineST/&#39;)</span>
<span class="c1"># import FineST as fst</span>
<span class="c1"># from FineST.datasets import dataset</span>
<span class="c1"># import FineST.plottings as fstplt</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="output_javascript"></div>
<script type="text/javascript">
var element = document.currentScript.previousSibling.previousSibling;
(function(root) {
  function now() {
    return new Date();
  }

  var force = true;
  var py_version = '3.1.1'.replace('rc', '-rc.').replace('.dev', '-dev.');
  var is_dev = py_version.indexOf("+") !== -1 || py_version.indexOf("-") !== -1;
  var reloading = false;
  var Bokeh = root.Bokeh;
  var bokeh_loaded = Bokeh != null && (Bokeh.version === py_version || (Bokeh.versions !== undefined && Bokeh.versions.has(py_version)));

  if (typeof (root._bokeh_timeout) === "undefined" || force) {
    root._bokeh_timeout = Date.now() + 5000;
    root._bokeh_failed_load = false;
  }

  function run_callbacks() {
    try {
      root._bokeh_onload_callbacks.forEach(function(callback) {
        if (callback != null)
          callback();
      });
    } finally {
      delete root._bokeh_onload_callbacks;
    }
    console.debug("Bokeh: all callbacks have finished");
  }

  function load_libs(css_urls, js_urls, js_modules, js_exports, callback) {
    if (css_urls == null) css_urls = [];
    if (js_urls == null) js_urls = [];
    if (js_modules == null) js_modules = [];
    if (js_exports == null) js_exports = {};

    root._bokeh_onload_callbacks.push(callback);

    if (root._bokeh_is_loading > 0) {
      console.debug("Bokeh: BokehJS is being loaded, scheduling callback at", now());
      return null;
    }
    if (js_urls.length === 0 && js_modules.length === 0 && Object.keys(js_exports).length === 0) {
      run_callbacks();
      return null;
    }
    if (!reloading) {
      console.debug("Bokeh: BokehJS not loaded, scheduling load and callback at", now());
    }

    function on_load() {
      root._bokeh_is_loading--;
      if (root._bokeh_is_loading === 0) {
        console.debug("Bokeh: all BokehJS libraries/stylesheets loaded");
        run_callbacks()
      }
    }
    window._bokeh_on_load = on_load

    function on_error() {
      console.error("failed to load " + url);
    }

    var skip = [];
    if (window.requirejs) {
      window.requirejs.config({'packages': {}, 'paths': {'jspanel': 'https://cdn.jsdelivr.net/npm/jspanel4@4.12.0/dist/jspanel', 'jspanel-modal': 'https://cdn.jsdelivr.net/npm/jspanel4@4.12.0/dist/extensions/modal/jspanel.modal', 'jspanel-tooltip': 'https://cdn.jsdelivr.net/npm/jspanel4@4.12.0/dist/extensions/tooltip/jspanel.tooltip', 'jspanel-hint': 'https://cdn.jsdelivr.net/npm/jspanel4@4.12.0/dist/extensions/hint/jspanel.hint', 'jspanel-layout': 'https://cdn.jsdelivr.net/npm/jspanel4@4.12.0/dist/extensions/layout/jspanel.layout', 'jspanel-contextmenu': 'https://cdn.jsdelivr.net/npm/jspanel4@4.12.0/dist/extensions/contextmenu/jspanel.contextmenu', 'jspanel-dock': 'https://cdn.jsdelivr.net/npm/jspanel4@4.12.0/dist/extensions/dock/jspanel.dock', 'gridstack': 'https://cdn.jsdelivr.net/npm/gridstack@7.2.3/dist/gridstack-all', 'notyf': 'https://cdn.jsdelivr.net/npm/notyf@3/notyf.min'}, 'shim': {'jspanel': {'exports': 'jsPanel'}, 'gridstack': {'exports': 'GridStack'}}});
      require(["jspanel"], function(jsPanel) {
    window.jsPanel = jsPanel
    on_load()
      })
      require(["jspanel-modal"], function() {
    on_load()
      })
      require(["jspanel-tooltip"], function() {
    on_load()
      })
      require(["jspanel-hint"], function() {
    on_load()
      })
      require(["jspanel-layout"], function() {
    on_load()
      })
      require(["jspanel-contextmenu"], function() {
    on_load()
      })
      require(["jspanel-dock"], function() {
    on_load()
      })
      require(["gridstack"], function(GridStack) {
    window.GridStack = GridStack
    on_load()
      })
      require(["notyf"], function() {
    on_load()
      })
      root._bokeh_is_loading = css_urls.length + 9;
    } else {
      root._bokeh_is_loading = css_urls.length + js_urls.length + js_modules.length + Object.keys(js_exports).length;
    }

    var existing_stylesheets = []
    var links = document.getElementsByTagName('link')
    for (var i = 0; i < links.length; i++) {
      var link = links[i]
      if (link.href != null) {
    existing_stylesheets.push(link.href)
      }
    }
    for (var i = 0; i < css_urls.length; i++) {
      var url = css_urls[i];
      if (existing_stylesheets.indexOf(url) !== -1) {
    on_load()
    continue;
      }
      const element = document.createElement("link");
      element.onload = on_load;
      element.onerror = on_error;
      element.rel = "stylesheet";
      element.type = "text/css";
      element.href = url;
      console.debug("Bokeh: injecting link tag for BokehJS stylesheet: ", url);
      document.body.appendChild(element);
    }    if (((window['jsPanel'] !== undefined) && (!(window['jsPanel'] instanceof HTMLElement))) || window.requirejs) {
      var urls = ['https://cdn.holoviz.org/panel/1.2.3/dist/bundled/floatpanel/jspanel4@4.12.0/dist/jspanel.js', 'https://cdn.holoviz.org/panel/1.2.3/dist/bundled/floatpanel/jspanel4@4.12.0/dist/extensions/modal/jspanel.modal.js', 'https://cdn.holoviz.org/panel/1.2.3/dist/bundled/floatpanel/jspanel4@4.12.0/dist/extensions/tooltip/jspanel.tooltip.js', 'https://cdn.holoviz.org/panel/1.2.3/dist/bundled/floatpanel/jspanel4@4.12.0/dist/extensions/hint/jspanel.hint.js', 'https://cdn.holoviz.org/panel/1.2.3/dist/bundled/floatpanel/jspanel4@4.12.0/dist/extensions/layout/jspanel.layout.js', 'https://cdn.holoviz.org/panel/1.2.3/dist/bundled/floatpanel/jspanel4@4.12.0/dist/extensions/contextmenu/jspanel.contextmenu.js', 'https://cdn.holoviz.org/panel/1.2.3/dist/bundled/floatpanel/jspanel4@4.12.0/dist/extensions/dock/jspanel.dock.js'];
      for (var i = 0; i < urls.length; i++) {
        skip.push(urls[i])
      }
    }    if (((window['GridStack'] !== undefined) && (!(window['GridStack'] instanceof HTMLElement))) || window.requirejs) {
      var urls = ['https://cdn.holoviz.org/panel/1.2.3/dist/bundled/gridstack/gridstack@7.2.3/dist/gridstack-all.js'];
      for (var i = 0; i < urls.length; i++) {
        skip.push(urls[i])
      }
    }    if (((window['Notyf'] !== undefined) && (!(window['Notyf'] instanceof HTMLElement))) || window.requirejs) {
      var urls = ['https://cdn.holoviz.org/panel/1.2.3/dist/bundled/notificationarea/notyf@3/notyf.min.js'];
      for (var i = 0; i < urls.length; i++) {
        skip.push(urls[i])
      }
    }    var existing_scripts = []
    var scripts = document.getElementsByTagName('script')
    for (var i = 0; i < scripts.length; i++) {
      var script = scripts[i]
      if (script.src != null) {
    existing_scripts.push(script.src)
      }
    }
    for (var i = 0; i < js_urls.length; i++) {
      var url = js_urls[i];
      if (skip.indexOf(url) !== -1 || existing_scripts.indexOf(url) !== -1) {
    if (!window.requirejs) {
      on_load();
    }
    continue;
      }
      var element = document.createElement('script');
      element.onload = on_load;
      element.onerror = on_error;
      element.async = false;
      element.src = url;
      console.debug("Bokeh: injecting script tag for BokehJS library: ", url);
      document.head.appendChild(element);
    }
    for (var i = 0; i < js_modules.length; i++) {
      var url = js_modules[i];
      if (skip.indexOf(url) !== -1 || existing_scripts.indexOf(url) !== -1) {
    if (!window.requirejs) {
      on_load();
    }
    continue;
      }
      var element = document.createElement('script');
      element.onload = on_load;
      element.onerror = on_error;
      element.async = false;
      element.src = url;
      element.type = "module";
      console.debug("Bokeh: injecting script tag for BokehJS library: ", url);
      document.head.appendChild(element);
    }
    for (const name in js_exports) {
      var url = js_exports[name];
      if (skip.indexOf(url) >= 0 || root[name] != null) {
    if (!window.requirejs) {
      on_load();
    }
    continue;
      }
      var element = document.createElement('script');
      element.onerror = on_error;
      element.async = false;
      element.type = "module";
      console.debug("Bokeh: injecting script tag for BokehJS library: ", url);
      element.textContent = `
      import ${name} from "${url}"
      window.${name} = ${name}
      window._bokeh_on_load()
      `
      document.head.appendChild(element);
    }
    if (!js_urls.length && !js_modules.length) {
      on_load()
    }
  };

  function inject_raw_css(css) {
    const element = document.createElement("style");
    element.appendChild(document.createTextNode(css));
    document.body.appendChild(element);
  }

  var js_urls = ["https://cdn.bokeh.org/bokeh/release/bokeh-3.1.1.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.1.1.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.1.1.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.1.1.min.js", "https://cdn.holoviz.org/panel/1.2.3/dist/panel.min.js"];
  var js_modules = [];
  var js_exports = {};
  var css_urls = [];
  var inline_js = [    function(Bokeh) {
      Bokeh.set_log_level("info");
    },
function(Bokeh) {} // ensure no trailing comma for IE
  ];

  function run_inline_js() {
    if ((root.Bokeh !== undefined) || (force === true)) {
      for (var i = 0; i < inline_js.length; i++) {
        inline_js[i].call(root, root.Bokeh);
      }
      // Cache old bokeh versions
      if (Bokeh != undefined && !reloading) {
    var NewBokeh = root.Bokeh;
    if (Bokeh.versions === undefined) {
      Bokeh.versions = new Map();
    }
    if (NewBokeh.version !== Bokeh.version) {
      Bokeh.versions.set(NewBokeh.version, NewBokeh)
    }
    root.Bokeh = Bokeh;
      }} else if (Date.now() < root._bokeh_timeout) {
      setTimeout(run_inline_js, 100);
    } else if (!root._bokeh_failed_load) {
      console.log("Bokeh: BokehJS failed to load within specified timeout.");
      root._bokeh_failed_load = true;
    }
    root._bokeh_is_initializing = false
  }

  function load_or_wait() {
    // Implement a backoff loop that tries to ensure we do not load multiple
    // versions of Bokeh and its dependencies at the same time.
    // In recent versions we use the root._bokeh_is_initializing flag
    // to determine whether there is an ongoing attempt to initialize
    // bokeh, however for backward compatibility we also try to ensure
    // that we do not start loading a newer (Panel>=1.0 and Bokeh>3) version
    // before older versions are fully initialized.
    if (root._bokeh_is_initializing && Date.now() > root._bokeh_timeout) {
      root._bokeh_is_initializing = false;
      root._bokeh_onload_callbacks = undefined;
      console.log("Bokeh: BokehJS was loaded multiple times but one version failed to initialize.");
      load_or_wait();
    } else if (root._bokeh_is_initializing || (typeof root._bokeh_is_initializing === "undefined" && root._bokeh_onload_callbacks !== undefined)) {
      setTimeout(load_or_wait, 100);
    } else {
      Bokeh = root.Bokeh;
      bokeh_loaded = Bokeh != null && (Bokeh.version === py_version || (Bokeh.versions !== undefined && Bokeh.versions.has(py_version)));
      root._bokeh_is_initializing = true
      root._bokeh_onload_callbacks = []
      if (!reloading && (!bokeh_loaded || is_dev)) {
    root.Bokeh = undefined;
      }
      load_libs(css_urls, js_urls, js_modules, js_exports, function() {
    console.debug("Bokeh: BokehJS plotting callback run at", now());
    run_inline_js();
      });
    }
  }
  // Give older versions of the autoload script a head-start to ensure
  // they initialize before we start loading newer version.
  setTimeout(load_or_wait, 100)
}(window));
</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="output_javascript"></div>
<script type="text/javascript">
var element = document.currentScript.previousSibling.previousSibling;

if ((window.PyViz === undefined) || (window.PyViz instanceof HTMLElement)) {
  window.PyViz = {comms: {}, comm_status:{}, kernels:{}, receivers: {}, plot_index: []}
}


    function JupyterCommManager() {
    }

    JupyterCommManager.prototype.register_target = function(plot_id, comm_id, msg_handler) {
      if (window.comm_manager || ((window.Jupyter !== undefined) && (Jupyter.notebook.kernel != null))) {
        var comm_manager = window.comm_manager || Jupyter.notebook.kernel.comm_manager;
        comm_manager.register_target(comm_id, function(comm) {
          comm.on_msg(msg_handler);
        });
      } else if ((plot_id in window.PyViz.kernels) && (window.PyViz.kernels[plot_id])) {
        window.PyViz.kernels[plot_id].registerCommTarget(comm_id, function(comm) {
          comm.onMsg = msg_handler;
        });
      } else if (typeof google != 'undefined' && google.colab.kernel != null) {
        google.colab.kernel.comms.registerTarget(comm_id, (comm) => {
          var messages = comm.messages[Symbol.asyncIterator]();
          function processIteratorResult(result) {
            var message = result.value;
            console.log(message)
            var content = {data: message.data, comm_id};
            var buffers = []
            for (var buffer of message.buffers || []) {
              buffers.push(new DataView(buffer))
            }
            var metadata = message.metadata || {};
            var msg = {content, buffers, metadata}
            msg_handler(msg);
            return messages.next().then(processIteratorResult);
          }
          return messages.next().then(processIteratorResult);
        })
      }
    }

    JupyterCommManager.prototype.get_client_comm = function(plot_id, comm_id, msg_handler) {
      if (comm_id in window.PyViz.comms) {
        return window.PyViz.comms[comm_id];
      } else if (window.comm_manager || ((window.Jupyter !== undefined) && (Jupyter.notebook.kernel != null))) {
        var comm_manager = window.comm_manager || Jupyter.notebook.kernel.comm_manager;
        var comm = comm_manager.new_comm(comm_id, {}, {}, {}, comm_id);
        if (msg_handler) {
          comm.on_msg(msg_handler);
        }
      } else if ((plot_id in window.PyViz.kernels) && (window.PyViz.kernels[plot_id])) {
        var comm = window.PyViz.kernels[plot_id].connectToComm(comm_id);
        comm.open();
        if (msg_handler) {
          comm.onMsg = msg_handler;
        }
      } else if (typeof google != 'undefined' && google.colab.kernel != null) {
        var comm_promise = google.colab.kernel.comms.open(comm_id)
        comm_promise.then((comm) => {
          window.PyViz.comms[comm_id] = comm;
          if (msg_handler) {
            var messages = comm.messages[Symbol.asyncIterator]();
            function processIteratorResult(result) {
              var message = result.value;
              var content = {data: message.data};
              var metadata = message.metadata || {comm_id};
              var msg = {content, metadata}
              msg_handler(msg);
              return messages.next().then(processIteratorResult);
            }
            return messages.next().then(processIteratorResult);
          }
        })
        var sendClosure = (data, metadata, buffers, disposeOnDone) => {
          return comm_promise.then((comm) => {
            comm.send(data, metadata, buffers, disposeOnDone);
          });
        };
        var comm = {
          send: sendClosure
        };
      }
      window.PyViz.comms[comm_id] = comm;
      return comm;
    }
    window.PyViz.comm_manager = new JupyterCommManager();



var JS_MIME_TYPE = 'application/javascript';
var HTML_MIME_TYPE = 'text/html';
var EXEC_MIME_TYPE = 'application/vnd.holoviews_exec.v0+json';
var CLASS_NAME = 'output';

/**
 * Render data to the DOM node
 */
function render(props, node) {
  var div = document.createElement("div");
  var script = document.createElement("script");
  node.appendChild(div);
  node.appendChild(script);
}

/**
 * Handle when a new output is added
 */
function handle_add_output(event, handle) {
  var output_area = handle.output_area;
  var output = handle.output;
  if ((output.data == undefined) || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {
    return
  }
  var id = output.metadata[EXEC_MIME_TYPE]["id"];
  var toinsert = output_area.element.find("." + CLASS_NAME.split(' ')[0]);
  if (id !== undefined) {
    var nchildren = toinsert.length;
    var html_node = toinsert[nchildren-1].children[0];
    html_node.innerHTML = output.data[HTML_MIME_TYPE];
    var scripts = [];
    var nodelist = html_node.querySelectorAll("script");
    for (var i in nodelist) {
      if (nodelist.hasOwnProperty(i)) {
        scripts.push(nodelist[i])
      }
    }

    scripts.forEach( function (oldScript) {
      var newScript = document.createElement("script");
      var attrs = [];
      var nodemap = oldScript.attributes;
      for (var j in nodemap) {
        if (nodemap.hasOwnProperty(j)) {
          attrs.push(nodemap[j])
        }
      }
      attrs.forEach(function(attr) { newScript.setAttribute(attr.name, attr.value) });
      newScript.appendChild(document.createTextNode(oldScript.innerHTML));
      oldScript.parentNode.replaceChild(newScript, oldScript);
    });
    if (JS_MIME_TYPE in output.data) {
      toinsert[nchildren-1].children[1].textContent = output.data[JS_MIME_TYPE];
    }
    output_area._hv_plot_id = id;
    if ((window.Bokeh !== undefined) && (id in Bokeh.index)) {
      window.PyViz.plot_index[id] = Bokeh.index[id];
    } else {
      window.PyViz.plot_index[id] = null;
    }
  } else if (output.metadata[EXEC_MIME_TYPE]["server_id"] !== undefined) {
    var bk_div = document.createElement("div");
    bk_div.innerHTML = output.data[HTML_MIME_TYPE];
    var script_attrs = bk_div.children[0].attributes;
    for (var i = 0; i < script_attrs.length; i++) {
      toinsert[toinsert.length - 1].childNodes[1].setAttribute(script_attrs[i].name, script_attrs[i].value);
    }
    // store reference to server id on output_area
    output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE]["server_id"];
  }
}

/**
 * Handle when an output is cleared or removed
 */
function handle_clear_output(event, handle) {
  var id = handle.cell.output_area._hv_plot_id;
  var server_id = handle.cell.output_area._bokeh_server_id;
  if (((id === undefined) || !(id in PyViz.plot_index)) && (server_id !== undefined)) { return; }
  var comm = window.PyViz.comm_manager.get_client_comm("hv-extension-comm", "hv-extension-comm", function () {});
  if (server_id !== null) {
    comm.send({event_type: 'server_delete', 'id': server_id});
    return;
  } else if (comm !== null) {
    comm.send({event_type: 'delete', 'id': id});
  }
  delete PyViz.plot_index[id];
  if ((window.Bokeh !== undefined) & (id in window.Bokeh.index)) {
    var doc = window.Bokeh.index[id].model.document
    doc.clear();
    const i = window.Bokeh.documents.indexOf(doc);
    if (i > -1) {
      window.Bokeh.documents.splice(i, 1);
    }
  }
}

/**
 * Handle kernel restart event
 */
function handle_kernel_cleanup(event, handle) {
  delete PyViz.comms["hv-extension-comm"];
  window.PyViz.plot_index = {}
}

/**
 * Handle update_display_data messages
 */
function handle_update_output(event, handle) {
  handle_clear_output(event, {cell: {output_area: handle.output_area}})
  handle_add_output(event, handle)
}

function register_renderer(events, OutputArea) {
  function append_mime(data, metadata, element) {
    // create a DOM node to render to
    var toinsert = this.create_output_subarea(
    metadata,
    CLASS_NAME,
    EXEC_MIME_TYPE
    );
    this.keyboard_manager.register_events(toinsert);
    // Render to node
    var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};
    render(props, toinsert[0]);
    element.append(toinsert);
    return toinsert
  }

  events.on('output_added.OutputArea', handle_add_output);
  events.on('output_updated.OutputArea', handle_update_output);
  events.on('clear_output.CodeCell', handle_clear_output);
  events.on('delete.Cell', handle_clear_output);
  events.on('kernel_ready.Kernel', handle_kernel_cleanup);

  OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {
    safe: true,
    index: 0
  });
}

if (window.Jupyter !== undefined) {
  try {
    var events = require('base/js/events');
    var OutputArea = require('notebook/js/outputarea').OutputArea;
    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {
      register_renderer(events, OutputArea);
    }
  } catch(err) {
  }
}

</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<style>*[data-root-id],
*[data-root-id] > * {
  box-sizing: border-box;
  font-family: var(--jp-ui-font-family);
  font-size: var(--jp-ui-font-size1);
  color: var(--vscode-editor-foreground, var(--jp-ui-font-color1));
}

/* Override VSCode background color */
.cell-output-ipywidget-background:has(
    > .cell-output-ipywidget-background > .lm-Widget > *[data-root-id]
  ),
.cell-output-ipywidget-background:has(> .lm-Widget > *[data-root-id]) {
  background-color: transparent !important;
}
</style></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div class="logo-block">
<img src='data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAYAAACqaXHeAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz
AAAB+wAAAfsBxc2miwAAABl0RVh0U29mdHdhcmUAd3d3Lmlua3NjYXBlLm9yZ5vuPBoAAA6zSURB
VHic7ZtpeFRVmsf/5966taWqUlUJ2UioBBJiIBAwCZtog9IOgjqACsogKtqirT2ttt069nQ/zDzt
tI4+CrJIREFaFgWhBXpUNhHZQoKBkIUASchWla1S+3ar7r1nPkDaCAnZKoQP/D7mnPOe9/xy76n3
nFSAW9ziFoPFNED2LLK5wcyBDObkb8ZkxuaoSYlI6ZcOKq1eWFdedqNzGHQBk9RMEwFAASkk0Xw3
ETacDNi2vtvc7L0ROdw0AjoSotQVkKSvHQz/wRO1lScGModBFbDMaNRN1A4tUBCS3lk7BWhQkgpD
lG4852/+7DWr1R3uHAZVQDsbh6ZPN7CyxUrCzJMRouusj0ipRwD2uKm0Zn5d2dFwzX1TCGhnmdGo
G62Nna+isiUqhkzuKrkQaJlPEv5mFl2fvGg2t/VnzkEV8F5ioioOEWkLG86fvbpthynjdhXYZziQ
x1hC9J2NFyi8vCTt91Fh04KGip0AaG9zuCk2wQCVyoNU3Hjezee9bq92duzzTmxsRJoy+jEZZZYo
GTKJ6SJngdJqAfRzpze0+jHreUtPc7gpBLQnIYK6BYp/uGhw9YK688eu7v95ysgshcg9qSLMo3JC
4jqLKQFBgdKDPoQ+Pltb8dUyQLpeDjeVgI6EgLIQFT5tEl3rn2losHVsexbZ3EyT9wE1uGdkIPcy
BGxn8QUq1QrA5nqW5i2tLqvrrM9NK6AdkVIvL9E9bZL/oyfMVd/jqvc8LylzRBKDJSzIExwhQzuL
QYGQj4rHfFTc8mUdu3E7yoLtbTe9gI4EqVgVkug2i5+uXGo919ixbRog+3fTbQ8qJe4ZOYNfMoTI
OoshUNosgO60AisX15aeI2PSIp5KiFLI9ubb1vV3Qb2ltwLakUCDAkWX7/nHKRmmGIl9VgYsUhJm
2NXjKYADtM1ygne9QQDIXlk49FBstMKx66D1v4+XuQr7vqTe0VcBHQlRWiOCbmmSYe2SqtL6q5rJ
zsTb7lKx3FKOYC4DoqyS/B5bvLPxvD9Qtf6saxYLQGJErmDOdOMr/zo96km1nElr8bmPOBwI9COv
HnFPRIwmkSOv9kcAS4heRsidOkpeWBgZM+UBrTFAXNYL5Vf2ii9c1trNzpYdaoVil3WIc+wdk+gQ
noie3ecCcxt9ITcLAPWt/laGEO/9U6PmzZkenTtsSMQ8uYywJVW+grCstAvCIaAdArAsIWkRDDs/
KzLm2YcjY1Lv0UdW73HabE9n6V66cxSzfEmuJssTpKGVp+0vHq73FwL46eOjpMpbRAnNmJFrGJNu
Ukf9Yrz+3rghiumCKNXXWPhLYcjxGsIpoCMsIRoFITkW8AuyM8jC1+/QLx4bozCEJIq38+1rtpR6
V/yzb8eBlRb3fo5l783N0CWolAzJHaVNzkrTzlEp2bQ2q3TC5gn6wpnoQAmwSiGh2GitnTmVMc5O
UyfKWUKCIsU7+fZDKwqdT6DDpvkzAX4/+AMFjk0tDp5GRXLpQ2MUmhgDp5gxQT8+Y7hyPsMi8uxF
71H0oebujHALECjFKaW9Lm68n18wXp2kVzIcABytD5iXFzg+WVXkegpAsOOYziqo0OkK76GyquC3
ltZAzMhhqlSNmmWTE5T6e3IN05ITFLM4GdN0vtZ3ob8Jh1NAKXFbm5PtLU/eqTSlGjkNAJjdgn/N
aedXa0tdi7+t9G0FIF49rtMSEgAs1kDLkTPO7ebm4IUWeyh1bKomXqlgMG6kJmHcSM0clYLJ8XtR
1GTnbV3F6I5wCGikAb402npp1h1s7LQUZZSMIfALFOuL3UUrfnS8+rez7v9qcold5tilgHbO1fjK
9ubb17u9oshxzMiUBKXWqJNxd+fqb0tLVs4lILFnK71H0Ind7uiPgACVcFJlrb0tV6DzxqqTIhUM
CwDf1/rrVhTa33/3pGPxJYdQ2l2cbgVcQSosdx8uqnDtbGjh9SlDVSMNWhlnilfqZk42Th2ZpLpf
xrHec5e815zrr0dfBZSwzkZfqsv+1FS1KUknUwPARVvItfKUY+cn57yP7qv07UE3p8B2uhUwLk09
e0SCOrK+hbdYHYLjRIl71wWzv9jpEoeOHhGRrJAzyEyNiJuUqX0g2sBN5kGK6y2Blp5M3lsB9Qh4
y2Ja6x6+i0ucmKgwMATwhSjdUu49tKrQ/pvN5d53ml2CGwCmJipmKjgmyuaXzNeL2a0AkQ01Th5j
2DktO3Jyk8f9vcOBQHV94OK+fPumJmvQHxJoWkaKWq9Vs+yUsbq0zGT1I4RgeH2b5wef7+c7bl8F
eKgoHVVZa8ZPEORzR6sT1BzDUAD/d9F78e2Tzv99v8D+fLVTqAKAsbGamKey1Mt9Ann4eH3gTXTz
idWtAJ8PQWOk7NzSeQn/OTHDuEikVF1R4z8BQCy+6D1aWRfY0tTGG2OM8rRoPaeIj5ZHzJxszElN
VM8K8JS5WOfv8mzRnQAKoEhmt8gyPM4lU9SmBK1MCQBnW4KONT86v1hZ1PbwSXPw4JWussVjtH9Y
NCoiL9UoH/6PSu8jFrfY2t36erQHXLIEakMi1SydmzB31h3GGXFDFNPaK8Rme9B79Ixrd0WN+1ij
NRQ/doRmuFLBkHSTOm5GruG+pFjFdAmorG4IXH1Qua6ASniclfFtDYt+oUjKipPrCQB7QBQ2lrgP
fFzm+9XWUtcqJ3/5vDLDpJ79XHZk3u8nGZ42qlj1+ydtbxysCezrydp6ugmipNJ7WBPB5tydY0jP
HaVNzs3QzeE4ZpTbI+ZbnSFPbVOw9vsfnVvqWnirPyCNGD08IlqtYkh2hjZ5dErEQzoNm+6ykyOt
Lt5/PQEuSRRKo22VkydK+vvS1XEKlhCJAnsqvcVvH7f/ZU2R67eXbMEGAMiIV5oWZWiWvz5Fv2xG
sjqNJQRvn3Rs2lji/lNP19VjAQDgD7FHhujZB9OGqYxRkZxixgRDVlqS6uEOFaJUVu0rPFzctrnF
JqijImVp8dEKVWyUXDk92zAuMZ6bFwpBU1HrOw6AdhQgUooChb0+ItMbWJitSo5Ws3IAOGEOtL53
0vHZih9sC4vtofZ7Qu6523V/fmGcds1TY3V36pUsBwAbSlxnVh2xLfAD/IAIMDf7XYIkNmXfpp2l
18rkAJAy9HKFaIr/qULkeQQKy9zf1JgDB2uaeFNGijo5QsUyacNUUTOnGO42xSnv4oOwpDi1zYkc
efUc3I5Gk6PhyTuVKaOGyLUAYPGIoY9Pu/atL/L92+4q9wbflRJ2Trpm/jPjdBtfnqB/dIThcl8A
KG7hbRuKnb8qsQsVvVlTrwQAQMUlf3kwJI24Z4JhPMtcfng5GcH49GsrxJpGvvHIaeem2ma+KSjQ
lIwUdYyCY8j4dE1KzijNnIP2llF2wcXNnsoapw9XxsgYAl6k+KzUXbi2yP3KR2ecf6z3BFsBICdW
nvnIaG3eHybqX7vbpEqUMT+9OL4Qpe8VON7dXuFd39v19FoAABRVePbGGuXTszO0P7tu6lghUonE
llRdrhArLvmKdh9u29jcFiRRkfLUxBiFNiqSU9icoZQHo5mYBI1MBgBH6wMNb+U7Pnw337H4gi1Y
ciWs+uks3Z9fztUvfzxTm9Ne8XXkvQLHNytOOZeiD4e0PgkAIAYCYknKUNUDSXEKzdWNpnil7r4p
xqkjTarZMtk/K8TQ6Qve78qqvXurGwIJqcOUKfUWHsm8KGvxSP68YudXq4pcj39X49uOK2X142O0
Tz5/u/7TVybqH0rSya6ZBwD21/gubbrgWdDgEOx9WUhfBaC2ibcEBYm7a7x+ukrBMNcEZggyR0TE
T8zUPjikQ4VosQZbTpS4vqizBKvqmvjsqnpfzaZyx9JPiz1/bfGKdgD45XB1zoIMzYbfTdS/NClB
Gct0USiY3YL/g0LHy/uq/Ef6uo5+n0R/vyhp17Klpge763f8rMu6YU/zrn2nml+2WtH+Z+5IAAFc
2bUTdTDOSNa9+cQY7YLsOIXhevEkCvzph7a8laecz/Un/z4/Ae04XeL3UQb57IwU9ZDr9UuKVajv
nxp1+1UVIo/LjztZkKH59fO3G/JemqCfmaCRqbqbd90ZZ8FfjtkfAyD0J/9+C2h1hDwsSxvGjNDc
b4zk5NfrSwiQblLHzZhg+Jf4aPlUwpDqkQqa9nimbt1/TDH8OitGMaQnj+RJS6B1fbF7SY1TqO5v
/v0WAADl1f7zokgS7s7VT2DZ7pegUjBM7mjtiDZbcN4j0YrHH0rXpCtY0qPX0cVL0rv5jv/ZXend
0u/EESYBAFBU4T4Qa5TflZOhTe7pmKpaP8kCVUVw1+yhXfJWvn1P3hnXi33JsTN6PnP3hHZ8Z3/h
aLHzmkNPuPj7Bc/F/Q38CwjTpSwQXgE4Vmwry9tpfq/ZFgqFMy4AVDtCvi8rvMvOmv0N4YwbVgEA
sPM72/KVnzfspmH7HQGCRLG2yL1+z8XwvPcdCbsAANh+xPzstgMtxeGKt+6MK3/tacfvwhWvIwMi
oKEBtm0H7W+UVfkc/Y1V0BhoPlDr/w1w/eu1vjIgAgDg22OtX6/eYfnEz/focrZTHAFR+PSs56/7
q32nwpjazxgwAQCwcU/T62t3WL7r6/jVRa6/byp1rei+Z98ZUAEAhEPHPc8fKnTU9nbgtnOe8h0l
9hcGIqmODLQAHCy2Xti6v/XNRivf43f4fFvIteu854+VHnR7q9tfBlwAAGz+pnndB9vM26UebAe8
SLHujPOTPVW+rwY+sxskAAC2HrA8t2Vvc7ffP1r9o+vwR2dcr92InIAbKKC1FZ5tB1tf+/G8p8sv
N/9Q5zd/XR34LYCwV5JdccMEAMDBk45DH243r/X4xGvqxFa/GNpS7n6rwOwNWwHVE26oAADYurf1
zx/utOzt+DMKYM0p17YtZZ5VNzqfsB2HewG1WXE8PoZ7gOclbTIvynZf9JV+fqZtfgs/8F/Nu5rB
EIBmJ+8QRMmpU7EzGRsf2FzuePqYRbzh/zE26EwdrT10f6r6o8HOYzCJB9Dpff8tbnGLG8L/A/WE
roTBs2RqAAAAAElFTkSuQmCC'
     style='height:25px; border-radius:12px; display: inline-block; float: left; vertical-align: middle'></img>


  <img src='data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACMAAAAjCAYAAAAe2bNZAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAK6wAACusBgosNWgAAABx0RVh0U29mdHdhcmUAQWRvYmUgRmlyZXdvcmtzIENTNui8sowAAAf9SURBVFiFvZh7cFTVHcc/59y7793sJiFAwkvAYDRqFWwdraLVlj61diRYsDjqCFbFKrYo0CltlSq1tLaC2GprGIriGwqjFu10OlrGv8RiK/IICYECSWBDkt3s695zTv9IAtlHeOn0O7Mzu797z+/3Ob/z+p0VfBq9doNFljuABwAXw2PcvGHt6bgwxhz7Ls4YZNVXxxANLENwE2D1W9PAGmAhszZ0/X9gll5yCbHoOirLzmaQs0F6F8QMZq1v/8xgNm7DYwwjgXJLYL4witQ16+sv/U9HdDmV4WrKw6B06cZC/RMrM4MZ7xz61DAbtzEXmAvUAX4pMOVecg9/MFFu3j3Gz7gQBLygS2RGumBkL0cubiFRsR3LzVBV1UMk3IrW73PT9C2lYOwhQB4ClhX1AuKpjLcV27oEjyUpNUJCg1CvcejykWTCXyQgzic2HIIBjg3pS6+uRLKAhumZvD4U+tq0jTrgkVKQQtLekfTtxIPAkhTNF6G7kZm7aPp6M9myKVQEoaYaIhEQYvD781DML/RfBGNZXAl4irJiwBa07e/y7cQnBaJghIX6ENl2GR/fGCBoz6cm5qeyEqQA5ZYA5x5eeiV0Qph4gjFAUSwAr6QllQgcxS/Jm25Cr2Tmpsk03XI9NfI31FTZBEOgVOk51adqDBNPCNPSRlkiDXbBEwOU2WxH+I7itQZ62g56OjM33suq1YsZHVtGZSUI2QdyYgkgOthQNIF7BIGDnRAJgJSgj69cUx1gB8PkOGwL4E1gPrM27gIg7NlGKLQApc7BmEnAxP5g/rw4YqBrCDB5xHkw5rdR/1qTrN/hKNo6YUwVDNpFsnjYS8RbidBPcPXFP6R6yfExuOXmN4A3jv1+8ZUwgY9D2OWjUZE6lO88jDwHI8ZixGiMKSeYTBamCoDk6kDAb6y1OcH1a6KpD/fZesoFw5FlIXAVCIiH4PxrV+p2npVDToTBmtjY8t1swh2V61E9KqWiyuPEjM8dbfxuvfa49Zayf9R136Wr8mBSf/T7bNteA8zwaGEUbFpckWwq95n59dUIywKl2fbOIS5e8bWSu0tJ1a5redAYfqkdjesodFajcgaVNWhXo1C9SrkN3Usmv3UMJrc6/DDwkwEntkEJLe67tSLhvyzK8rHDQWleve5CGk4VZEB1r+5bg2E2si+Y0QatDK6jUVkX5eg2YYlp++ZM+rfMNYamAj8Y7MAVWFqaR1f/t2xzU4IHjybBtthzuiAASqv7jTF7jOqDMAakFHgDNsFyP+FhwZHBmH9F7cutIYkQCylYYv1AZSqsn1/+bX51OMMjPSl2nAnM7hnjOx2v53YgNWAzHM9Q/9l0lQWPSCBSyokAtOBC1Rj+w/1Xs+STDp4/E5g7Rs2zm2+oeVd7PUuHKDf6A4r5EsPT5K3gfCnBXNUYnvGzb+KcCczYYWOnLpy4eOXuG2oec0PBN8XQQAnpvS35AvAykr56rWhPBiV4MvtceGLxk5Mr6A1O8IfK7rl7xJ0r9kyumuP4fa0lMqTBLJIAJqEf1J3qE92lMBndlyfRD2YBghHC4hlny7ASqCeWo5zaoDdIWfnIefNGTb9fC73QDfhyBUCNOxrGPSUBfPem9us253YTV+3mcBbdkUYfzmHiLqZbYdIGHHON2ZlemXouaJUOO6TqtdHEQuXYY8Yt+EbDgmlS6RdzkaDTv2P9A3gICiq93sWhb5mc5wVhuU3Y7m5hOc3So7qFT3SLgOXHb/cyOfMn7xROegoC/PTcn3v8gbKPgDopJFk3R/uBPWQiwQ+2/GJevRMObLUzqe/saJjQUQTTftEVMW9tWxPgAocwcj9abNcZe7s+6t2R2xXZG7zyYLp8Q1PiRBBHym5bYuXi8Qt+/LvGu9f/5YDAxABsaRNPH6Xr4D4Sk87a897SOy9v/fKwjoF2eQel95yDESGEF6gEMwKhLwKus3wOVjTtes7qzgLdXTMnNCNoEpbcrtNuq6N7Xh/+eqcbj94xQkp7mdKpW5XbtbR8Z26kgMCAf2UU5YEovRUVRHbu2b3vK1UdDFkDCyMRQxbpdv8nhKAGIa7QaQedzT07fFPny53R738JoVYBdVrnsNx9XZ9v33UeGO+AA2MMUkgqQ5UcdDLZSFeVgONnXeHqSAC5Ew1BXwko0D1Zct3dT1duOjS3MzZnEUJtBuoQAq3SGOLR4ekjn9NC5nVOaYXf9lETrUkmOJy3pOz8OKIb2A1cWhJCCEzOxU2mUPror+2/L3yyM3pkM7jTjr1nBOgkGeyQ7erxpdJsMAS9wb2F9rzMxNY1K2PMU0WtZV82VU8Wp6vbKJVo9Lx/+4cydORdxCCQ/kDGTZCWsRpLu7VD7bfKqL8V2orKTp/PtzaXy42jr6TwAuisi+7JolUG4wY+8vyrISCMtRrLKWpvjAOqx/QGhp0rjRo5xD3x98CWQuOQN8qumRMmI7jKZPUEpzNVZsj4Zbaq1to5tZZsKIydLWojhIXrJnES79EaOzv3du2NytKuxzJKAA6wF8xqEE8s2jo/1wd/khslQGxd81Zg62Bbp31XBH+iETt7Y3ELA0iU6iGDlQ5mexe0VEx4a3x8V1AaYwFJgTiwaOsDmeK2J8nMUOqsnB1A+dcA04ucCYt0urkjmflk9iT2v30q/gZn5rQPvor4n9Ou634PeBzoznes/iot/7WnClKoM/+zCIjH5kwT8ChQjTHPIPTjFV3PpU/Hx+DM/A9U3IXI4SPCYAAAAABJRU5ErkJggg=='
       style='height:15px; border-radius:12px; display: inline-block; float: left'></img>





</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">path</span> <span class="o">=</span> <span class="s1">&#39;/mnt/lingyu/nfs_share2/Python/&#39;</span>

<span class="n">fst</span><span class="o">.</span><span class="n">setup_seed</span><span class="p">(</span><span class="mi">666</span><span class="p">)</span>

<span class="k">if</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">is_available</span><span class="p">():</span>
    <span class="n">dev</span> <span class="o">=</span> <span class="s2">&quot;cuda:0&quot;</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">dev</span> <span class="o">=</span> <span class="s2">&quot;cpu&quot;</span>
<span class="n">device</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">device</span><span class="p">(</span><span class="n">dev</span><span class="p">)</span>

<span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;#000003&quot;</span><span class="p">,</span>  <span class="s2">&quot;#3b0f6f&quot;</span><span class="p">,</span>  <span class="s2">&quot;#8c2980&quot;</span><span class="p">,</span>   <span class="s2">&quot;#f66e5b&quot;</span><span class="p">,</span> <span class="s2">&quot;#fd9f6c&quot;</span><span class="p">,</span> <span class="s2">&quot;#fbfcbf&quot;</span><span class="p">]</span>
<span class="n">cnt_color</span> <span class="o">=</span> <span class="n">clr</span><span class="o">.</span><span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span><span class="s1">&#39;magma&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="p">,</span> <span class="n">N</span><span class="o">=</span><span class="mi">256</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>0.1 Make logging and save model</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">()</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>

<span class="n">model_folder</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;FineST/FineST_local/Finetune/&#39;</span>
<span class="n">dir_name</span> <span class="o">=</span> <span class="n">model_folder</span> <span class="o">+</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s1">&#39;%Y%m</span><span class="si">%d</span><span class="s1">%H%M%S</span><span class="si">%f</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">dir_name</span><span class="p">):</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">dir_name</span><span class="p">)</span>
    <span class="n">logger</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">setup_logger</span><span class="p">(</span><span class="n">dir_name</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">dir_name</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
/mnt/lingyu/nfs_share2/Python/FineST/FineST_local/Finetune/20241111220901614764
</pre></div></div>
</div>
<p>0.2 Load parameter settings</p>
<p>You can use the <code class="docutils literal notranslate"><span class="pre">parameter_file_path</span></code> you set or the optimal parameters obtained from your train and test model for your dataset, but here we use our trained <code class="docutils literal notranslate"><span class="pre">parameter_file_path</span></code> for paper results repeated (saved in <code class="docutils literal notranslate"><span class="pre">FineST/FineST/Parameter/parameters_NPC_P10125.json</span></code>).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">parameter_file_path</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;FineST/FineST/Parameter/parameters_NPC_P10125.json&#39;</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">parameter_file_path</span><span class="p">,</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">json_file</span><span class="p">:</span>
    <span class="n">params</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">json_file</span><span class="p">)</span>

<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Load parameters:</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">2</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
[2024-11-11 22:09:01] INFO - Load parameters:
{
  &#34;n_encoder_hidden_matrix&#34;: 256,
  &#34;n_encoder_hidden_image&#34;: 128,
  &#34;n_encoder_latent&#34;: 128,
  &#34;n_projection_hidden&#34;: 256,
  &#34;n_projection_output&#34;: 128,
  &#34;batch_size&#34;: 200,
  &#34;batch_size_pair&#34;: 64,
  &#34;n_encoder_layers&#34;: 2,
  &#34;dropout_rate&#34;: 0,
  &#34;training_epoch&#34;: 50,
  &#34;inital_learning_rate&#34;: 0.1,
  &#34;k_nearest_positives&#34;: 0,
  &#34;temperature&#34;: 0.03
}
</pre></div></div>
</div>
</section>
<section id="1.-Load-ST-data-and-Image-embeddings">
<h2>1. Load ST data and Image embeddings<a class="headerlink" href="#1.-Load-ST-data-and-Image-embeddings" title="Permalink to this heading"></a></h2>
<p>1.1 Load spatial data</p>
<p>A human Nasopharyngeal carcinoma dataset (NPC_patient_1) from <a class="reference external" href="https://www.nature.com/articles/s41467-023-37614-6">Gong, et al</a>. will be used as a reference dataset, raw_feature_bc_matrix, filtered_feature_bc_matrix, and spatial coordinates (.h5) are publicly available in <a class="reference external" href="https://ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE200310">GSE200310</a>. For easier reuse, we included them in an anndata object which can be loaded directly in FineST Python package.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">NPC</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 1331 × 36601
    obs: &#39;in_tissue&#39;, &#39;array_row&#39;, &#39;array_col&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;
    uns: &#39;spatial&#39;
    obsm: &#39;spatial&#39;
</pre></div></div>
</div>
<p>1.2 Selected LR genes</p>
<p>In FineST, we only considers 963 human genes involved in ligand-receptor pairs from <a class="reference external" href="https://github.com/sqjin/CellChat/tree/master/data">CellChatDB (version.1.1.3)</a> (mouse: 2,022 pairs, human: 1,940 pairs, zebrafish: 2,774 pairs), and extract the overlap gene’s expression profile.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">LR_gene_path</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;FineST/FineST/Dataset/LRgene/LRgene_CellChatDB_baseline.csv&#39;</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">adata_LR</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">LR_gene_path</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 1331 × 911
    obs: &#39;in_tissue&#39;, &#39;array_row&#39;, &#39;array_col&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;
    uns: &#39;spatial&#39;
    obsm: &#39;spatial&#39;
</pre></div></div>
</div>
<p>1.3 Data preprocess</p>
<p>After making an intersection with the LR genes and filtering out genes that are detected in less than 10 cells, 596 genes from the NPC datasetare used for training and prediction.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">adata_preprocess</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

<span class="n">gene_hv</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
<span class="n">matrix</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">adata2matrix</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">gene_hv</span><span class="p">)</span>
<span class="n">matrix</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 1331 × 596
    obs: &#39;in_tissue&#39;, &#39;array_row&#39;, &#39;array_col&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;total_counts_mt&#39;, &#39;log1p_total_counts_mt&#39;, &#39;pct_counts_mt&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;mt&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;
    uns: &#39;spatial&#39;, &#39;log1p&#39;
    obsm: &#39;spatial&#39;
(1331, 596)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>TGFB1</th>
      <th>TGFBR1</th>
      <th>TGFBR2</th>
      <th>TGFB2</th>
      <th>TGFB3</th>
      <th>ACVR1B</th>
      <th>ACVR1C</th>
      <th>ACVR1</th>
      <th>BMP2</th>
      <th>BMPR1A</th>
      <th>...</th>
      <th>KDR</th>
      <th>TREM2</th>
      <th>SEMA6A</th>
      <th>SEMA6B</th>
      <th>SEMA7A</th>
      <th>PLXNC1</th>
      <th>SIGLEC1</th>
      <th>THY1</th>
      <th>VCAM1</th>
      <th>VSIR</th>
    </tr>
    <tr>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAACAATCTACTAGCA-1</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.00000</td>
      <td>0.0</td>
      <td>0.784856</td>
      <td>0.784856</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.784856</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>AAACGAAGAACATACC-1</th>
      <td>1.538792</td>
      <td>0.000000</td>
      <td>0.00000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.797351</td>
      <td>0.000000</td>
      <td>0.0000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.797351</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.797351</td>
      <td>0.000000</td>
      <td>1.235270</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>AAAGGGATGTAGCAAG-1</th>
      <td>0.000000</td>
      <td>0.481666</td>
      <td>0.00000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.481666</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.481666</td>
      <td>1.409484</td>
      <td>0.481666</td>
    </tr>
    <tr>
      <th>AAAGTCACTGATGTAA-1</th>
      <td>0.000000</td>
      <td>0.633767</td>
      <td>0.00000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>1.018628</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.633767</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.633767</td>
      <td>0.000000</td>
      <td>1.512659</td>
      <td>0.633767</td>
    </tr>
    <tr>
      <th>AAAGTGTGATTTATCT-1</th>
      <td>0.501523</td>
      <td>0.000000</td>
      <td>1.08306</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.501523</td>
      <td>0.833982</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TTGTGGCCCTGACAGT-1</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.00000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.888100</td>
      <td>0.000000</td>
      <td>0.8881</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.888100</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.888100</td>
      <td>0.888100</td>
    </tr>
    <tr>
      <th>TTGTGTTTCCCGAAAG-1</th>
      <td>0.890100</td>
      <td>0.000000</td>
      <td>0.00000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.540979</td>
      <td>0.540979</td>
      <td>0.000000</td>
      <td>0.540979</td>
      <td>0.540979</td>
      <td>0.890100</td>
    </tr>
    <tr>
      <th>TTGTTAGCAAATTCGA-1</th>
      <td>0.457721</td>
      <td>0.457721</td>
      <td>0.00000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.457721</td>
      <td>0.0</td>
      <td>0.457721</td>
      <td>0.457721</td>
      <td>0.0000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.457721</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.500249</td>
      <td>0.457721</td>
    </tr>
    <tr>
      <th>TTGTTCAGTGTGCTAC-1</th>
      <td>1.879867</td>
      <td>0.000000</td>
      <td>0.00000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>2.493640</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>TTGTTTGTGTAAATTC-1</th>
      <td>0.888100</td>
      <td>0.000000</td>
      <td>0.00000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.888100</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.888100</td>
      <td>1.350931</td>
    </tr>
  </tbody>
</table>
<p>1331 rows × 596 columns</p>
</div></div>
</div>
<p>1.4 Order by image file name</p>
<p>Load image embedding from <strong>Setp0</strong> <code class="docutils literal notranslate"><span class="pre">HIPT_image_feature_extract.py</span></code>. For NPC dataset, the image embedding file can be downloaded <a class="reference external" href="https://drive.google.com/drive/folders/1ypaTohLM2lP4EnKeYTS6QUvv-ymcRVZo">Goole Drive</a> (within spots: AH_Patient1_pth_64_16, between spots: NEW_AH_Patient1_pth_64_16, single-nuclei: sc_Patient1_pth_16_16).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">file_paths</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;NPC/Data/stdata/ZhuoLiang/LLYtest/AH_Patient1_pth_64_16/&#39;</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Image embedding file:&quot;</span><span class="p">,</span> <span class="n">file_paths</span><span class="p">[:</span><span class="mi">3</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Image embedding file: [&#39;AH_Patient1_10014_10023.pth&#39;, &#39;AH_Patient1_10015_9649.pth&#39;, &#39;AH_Patient1_10016_9276.pth&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Image patch position</span>
<span class="n">position_image</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">get_image_coord</span><span class="p">(</span><span class="n">file_paths</span><span class="p">,</span> <span class="n">dataset_class</span><span class="o">=</span><span class="s2">&quot;Visium&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">position_image</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">position_image</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(1331, 2)
   pixel_x  pixel_y
0    10023    10014
1     9649    10015
2     9276    10016
3     7783    10020
4     7410    10021
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## ST spot position</span>
<span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;FineST/FineST/Dataset/NPC/patient1/&#39;</span><span class="p">)</span>
<span class="n">position</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;tissue_positions_list.csv&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">position</span> <span class="o">=</span> <span class="n">position</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="n">position</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">]:</span> <span class="s1">&#39;pixel_x&#39;</span><span class="p">,</span>
                                    <span class="n">position</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span> <span class="s1">&#39;pixel_y&#39;</span><span class="p">})</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## merge position</span>
<span class="n">position_image</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">image_coord_merge</span><span class="p">(</span><span class="n">position_image</span><span class="p">,</span> <span class="n">position</span><span class="p">,</span> <span class="n">dataset</span> <span class="o">=</span> <span class="s1">&#39;Visium&#39;</span><span class="p">)</span>
<span class="n">spotID_order</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">position_image</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">position_image</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">position_image</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(1331, 6)
                    0  1   x   y  pixel_x  pixel_y
0  GTCGTTATTCGCTTAT-1  1  46  76    10023    10014
1  ACAAGGGCAGGCTCTG-1  1  44  76     9649    10015
2  TGCGTTTGTTGACACT-1  1  42  76     9276    10016
3  TTGAATATGGACTTTC-1  1  34  76     7783    10020
4  CCGGGCGGTCTCGTCA-1  1  32  76     7410    10021
</pre></div></div>
</div>
<p>1.5 Order matrix row/barcode, spatial coordinates by image coordinates</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">matrix_order</span><span class="p">,</span> <span class="n">matrix_order_df</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">sort_matrix</span><span class="p">(</span><span class="n">matrix</span><span class="p">,</span> <span class="n">position_image</span><span class="p">,</span> <span class="n">spotID_order</span><span class="p">,</span> <span class="n">gene_hv</span><span class="p">)</span>
<span class="n">matrix_order_df</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>TGFB1</th>
      <th>TGFBR1</th>
      <th>TGFBR2</th>
      <th>TGFB2</th>
      <th>TGFB3</th>
      <th>ACVR1B</th>
      <th>ACVR1C</th>
      <th>ACVR1</th>
      <th>BMP2</th>
      <th>BMPR1A</th>
      <th>...</th>
      <th>KDR</th>
      <th>TREM2</th>
      <th>SEMA6A</th>
      <th>SEMA6B</th>
      <th>SEMA7A</th>
      <th>PLXNC1</th>
      <th>SIGLEC1</th>
      <th>THY1</th>
      <th>VCAM1</th>
      <th>VSIR</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GTCGTTATTCGCTTAT-1</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.463911</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.463911</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>ACAAGGGCAGGCTCTG-1</th>
      <td>1.543805</td>
      <td>0.905412</td>
      <td>0.905412</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.551857</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.551857</td>
      <td>0.551857</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.905412</td>
      <td>1.543805</td>
      <td>0.551857</td>
    </tr>
    <tr>
      <th>TGCGTTTGTTGACACT-1</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.092333</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>1.092333</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>TTGAATATGGACTTTC-1</th>
      <td>0.418937</td>
      <td>0.418937</td>
      <td>0.713288</td>
      <td>0.0</td>
      <td>0.418937</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.418937</td>
      <td>0.000000</td>
      <td>0.418937</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.713288</td>
      <td>0.940412</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>CCGGGCGGTCTCGTCA-1</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.848202</td>
      <td>0.848202</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>CCTCTGGCCTAGACGG-1</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.432448</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.432448</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.964348</td>
      <td>0.432448</td>
    </tr>
    <tr>
      <th>CAAACTATTGAGCTTC-1</th>
      <td>0.925633</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.925633</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.925633</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>ACATCGTTAACCTAGT-1</th>
      <td>0.682651</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.682651</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.682651</td>
      <td>1.084593</td>
      <td>1.084593</td>
    </tr>
    <tr>
      <th>GTTAGAGTGTGCCGCT-1</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.978859</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>1.250511</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>TCAGTTACGGAATGAT-1</th>
      <td>0.640733</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.640733</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.640733</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>1.306606</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
<p>1331 rows × 596 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## save gene expression mateix -- train model used</span>
<span class="c1"># np.save(str(path)+&#39;FineST/FineST_local/Dataset/NPC/ContrastP1geneLR/harmony_matrix.npy&#39;, matrix_order_df.T)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">update_adata_coord</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">matrix_order</span><span class="p">,</span> <span class="n">position_image</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 1331 × 596
    obs: &#39;in_tissue&#39;, &#39;array_row&#39;, &#39;array_col&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;total_counts_mt&#39;, &#39;log1p_total_counts_mt&#39;, &#39;pct_counts_mt&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;mt&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;
    uns: &#39;spatial&#39;, &#39;log1p&#39;
    obsm: &#39;spatial&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">position_order</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">update_st_coord</span><span class="p">(</span><span class="n">position_image</span><span class="p">)</span>
<span class="n">position_order</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pixel_y</th>
      <th>pixel_x</th>
      <th>array_row</th>
      <th>array_col</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>10014</td>
      <td>10023</td>
      <td>76</td>
      <td>46</td>
    </tr>
    <tr>
      <th>1</th>
      <td>10015</td>
      <td>9649</td>
      <td>76</td>
      <td>44</td>
    </tr>
    <tr>
      <th>2</th>
      <td>10016</td>
      <td>9276</td>
      <td>76</td>
      <td>42</td>
    </tr>
    <tr>
      <th>3</th>
      <td>10020</td>
      <td>7783</td>
      <td>76</td>
      <td>34</td>
    </tr>
    <tr>
      <th>4</th>
      <td>10021</td>
      <td>7410</td>
      <td>76</td>
      <td>32</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>1326</th>
      <td>9922</td>
      <td>4237</td>
      <td>75</td>
      <td>15</td>
    </tr>
    <tr>
      <th>1327</th>
      <td>9923</td>
      <td>3490</td>
      <td>75</td>
      <td>11</td>
    </tr>
    <tr>
      <th>1328</th>
      <td>9924</td>
      <td>3117</td>
      <td>75</td>
      <td>9</td>
    </tr>
    <tr>
      <th>1329</th>
      <td>9925</td>
      <td>2743</td>
      <td>75</td>
      <td>7</td>
    </tr>
    <tr>
      <th>1330</th>
      <td>9927</td>
      <td>1997</td>
      <td>75</td>
      <td>3</td>
    </tr>
  </tbody>
</table>
<p>1331 rows × 4 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## save the position data  -- train model used</span>
<span class="c1"># position_order.to_csv(str(path)+&#39;FineST/FineST_local/Dataset/NPC/ContrastP1geneLR/position_order.csv&#39;, index=False, header=False)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## save the original adata</span>
<span class="c1"># patientxy = &#39;patient1&#39;</span>
<span class="c1"># adata.write_h5ad(str(path)+&#39;FineST/FineST_local/Dataset/ImputData/&#39;+str(patientxy)+&#39;/&#39;+str(patientxy)+&#39;_adata_orignal.h5ad&#39;)</span>
</pre></div>
</div>
</div>
<div class="admonition note">
<p><strong>Note</strong> The above cells for <code class="docutils literal notranslate"><span class="pre">section</span> <span class="pre">1.5</span></code> only need to be run once. They aim to generate the ordered gene expression profile <code class="docutils literal notranslate"><span class="pre">harmony_matrix.npy</span></code> and the ordered ST spot coordinates <code class="docutils literal notranslate"><span class="pre">position_order.csv</span></code> according to image pixel coordinates, and save the normalized gene expression data <code class="docutils literal notranslate"><span class="pre">patient1_adata_orignal.h5ad</span></code>.</p>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## visualisation</span>
<span class="n">fstplt</span><span class="o">.</span><span class="n">gene_expr</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">matrix_order_df</span><span class="p">,</span> <span class="n">gene_selet</span><span class="o">=</span><span class="s1">&#39;CD70&#39;</span><span class="p">,</span> <span class="n">save_path</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/NPC_Train_Impute_34_0.png" src="_images/NPC_Train_Impute_34_0.png" />
</div>
</div>
</section>
<section id="2-Train-and-Test-model">
<h2>2 Train and Test model<a class="headerlink" href="#2-Train-and-Test-model" title="Permalink to this heading"></a></h2>
<p>2.1 Data loader and splitting: 80% for train and 20% for test</p>
<div class="line-block">
<div class="line">Input: three input file pathway:</div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">image_embed_path</span></code>: image embedding from <strong>Setp0</strong> <code class="docutils literal notranslate"><span class="pre">HIPT_image_feature_extract.py</span></code></div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">spatial_pos_path</span></code>: ordered ST spot coordinates, can be found in <code class="docutils literal notranslate"><span class="pre">FineST/Dataset/NPC/ContrastP1geneLR/position_order.csv</span></code></div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">reduced_mtx_path</span></code>: ordered gene expression profile, can be found in <code class="docutils literal notranslate"><span class="pre">FineST/Dataset/NPC/ContrastP1geneLR/harmony_matrix.np</span></code></div>
</div>
<p>For NPC dataset, the image embedding file can be downloaded <a class="reference external" href="https://drive.google.com/drive/folders/1ypaTohLM2lP4EnKeYTS6QUvv-ymcRVZo">Goole Drive</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">image_embed_path_NPC</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;NPC/Data/stdata/ZhuoLiang/LLYtest/AH_Patient1_pth_64_16/*.pth&#39;</span>
<span class="n">spatial_pos_path</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;FineST/FineST_local/Dataset/NPC/ContrastP1geneLR/position_order.csv&#39;</span>
<span class="n">reduced_mtx_path</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;FineST/FineST_local/Dataset/NPC/ContrastP1geneLR/harmony_matrix.npy&#39;</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">train_loader</span><span class="p">,</span> <span class="n">test_loader</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">build_loaders</span><span class="p">(</span><span class="n">batch_size</span><span class="o">=</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;batch_size&#39;</span><span class="p">],</span>
                                              <span class="n">image_embed_path</span><span class="o">=</span><span class="n">image_embed_path_NPC</span><span class="p">,</span>
                                              <span class="n">spatial_pos_path</span><span class="o">=</span><span class="n">spatial_pos_path</span><span class="p">,</span>
                                              <span class="n">reduced_mtx_path</span><span class="o">=</span><span class="n">reduced_mtx_path</span><span class="p">,</span>
                                              <span class="n">dataset_class</span><span class="o">=</span><span class="s1">&#39;Visium&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
***** Building loaders *****
Finished loading all files
train/test split completed
1064 267
***** Finished building loaders *****
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">all_dataset</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">build_loaders_inference</span><span class="p">(</span><span class="n">batch_size</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                          <span class="n">image_embed_path</span><span class="o">=</span><span class="n">image_embed_path_NPC</span><span class="p">,</span>
                                          <span class="n">spatial_pos_path</span><span class="o">=</span><span class="n">spatial_pos_path</span><span class="p">,</span>
                                          <span class="n">reduced_mtx_path</span><span class="o">=</span><span class="n">reduced_mtx_path</span><span class="p">,</span>
                                          <span class="n">dataset_class</span><span class="o">=</span><span class="s1">&#39;Visium&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
***** Building loaders_inference *****
Finished loading all files
***** Finished building loaders_inference *****
</pre></div></div>
</div>
<p>2.2 Train and test model on within spot</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># ## load parameter settings</span>
<span class="c1"># with open(parameter_file_path,&quot;r&quot;) as json_file:</span>
<span class="c1">#     params = json.load(json_file)</span>

<span class="c1"># ## add params</span>
<span class="c1"># params[&#39;n_input_matrix&#39;] = len(gene_hv)</span>
<span class="c1"># params[&#39;n_input_image&#39;] = 384</span>

<span class="c1"># ## init the model</span>
<span class="c1"># model = fst.CellContrastModel(n_input_matrix=params[&#39;n_input_matrix&#39;],</span>
<span class="c1">#                               n_input_image=params[&#39;n_input_image&#39;],</span>
<span class="c1">#                               n_encoder_hidden_matrix=params[&quot;n_encoder_hidden_matrix&quot;],</span>
<span class="c1">#                               n_encoder_hidden_image=params[&quot;n_encoder_hidden_image&quot;],</span>
<span class="c1">#                               n_encoder_latent=params[&quot;n_encoder_latent&quot;],</span>
<span class="c1">#                               n_projection_hidden=params[&quot;n_projection_hidden&quot;],</span>
<span class="c1">#                               n_projection_output=params[&quot;n_projection_output&quot;],</span>
<span class="c1">#                               n_encoder_layers=params[&quot;n_encoder_layers&quot;]).to(device)</span>

<span class="c1"># l = fst.ContrastiveLoss(temperature=params[&#39;temperature&#39;])</span>
<span class="c1"># # print(model)</span>

<span class="c1"># ## Set optimizer</span>
<span class="c1"># optimizer = torch.optim.SGD(model.parameters(), lr=params[&#39;inital_learning_rate&#39;],</span>
<span class="c1">#                             momentum=0.9, weight_decay=5e-4)</span>

<span class="c1"># ## Load the data</span>
<span class="c1"># train_loader, test_loader = fst.build_loaders(batch_size=params[&#39;batch_size&#39;],</span>
<span class="c1">#                                               image_embed_path=image_embed_path_NPC,</span>
<span class="c1">#                                               spatial_pos_path=spatial_pos_path,</span>
<span class="c1">#                                               reduced_mtx_path=reduced_mtx_path,</span>
<span class="c1">#                                               dataset_class=&#39;Visium&#39;)</span>


<span class="c1"># ## Train the model for a fixed number of epoch</span>
<span class="c1"># logger.info(&#39;Begin Training ...&#39;)</span>

<span class="c1"># start_time = time.time()</span>

<span class="c1"># best_loss = float(&#39;inf&#39;)</span>
<span class="c1"># best_epoch = 0</span>
<span class="c1"># for epoch in range(params[&#39;training_epoch&#39;]):</span>
<span class="c1">#     logger.info(&#39;epoch [{}/{}]&#39;.format(epoch + 1, epoch))</span>
<span class="c1">#     print(f&quot;Epoch: {epoch + 1}&quot;)</span>

<span class="c1">#     ######################################################################################</span>
<span class="c1">#     # Train the model</span>
<span class="c1">#     ######################################################################################</span>
<span class="c1">#     model.train()</span>
<span class="c1">#     start_time = time.time()</span>
<span class="c1">#     train_loss = fst.train_model(params, model, train_loader, optimizer, epoch, l,</span>
<span class="c1">#                                  tree_type=&#39;KDTree&#39;, leaf_size=2, dataset_class=&#39;Visium&#39;)</span>
<span class="c1">#     print(&quot;--- %s seconds ---&quot; % (time.time() - start_time))</span>

<span class="c1">#     ######################################################################################</span>
<span class="c1">#     # Evaluate the model</span>
<span class="c1">#     ######################################################################################</span>
<span class="c1">#     model.eval()</span>
<span class="c1">#     with torch.no_grad():</span>
<span class="c1">#         test_loss = fst.test_model(params, model, test_loader, l,</span>
<span class="c1">#                                    tree_type=&#39;KDTree&#39;, leaf_size=2, dataset_class=&#39;Visium&#39;)</span>

<span class="c1">#     if best_loss &gt; test_loss:</span>
<span class="c1">#         best_loss = test_loss</span>
<span class="c1">#         best_epoch = epoch</span>

<span class="c1">#         fst.save_model(model, dir_name, params, optimizer, train_loss)</span>
<span class="c1">#         print(&quot;Saved Best epoch &amp; Best Model! Loss: [{}: {}]&quot;.format(best_epoch, best_loss))</span>
<span class="c1">#         logger.info(&quot;Saved Best epoch &amp; Best Model! Loss: [{}: {}]&quot;.format(best_epoch, best_loss))</span>

<span class="c1"># print(&quot;Done!, final loss: {}&quot;.format(best_loss))</span>
<span class="c1"># print(&quot;Best epoch: {}&quot;.format(best_epoch))</span>

<span class="c1"># print(&quot;--- %s seconds ---&quot; % (time.time() - start_time))</span>

<span class="c1"># logger.info(&quot;Done!, Best epoch &amp; Best Model! Loss: [{}: {}]&quot;.format(best_epoch, best_loss))</span>

<span class="c1"># logger.info(&#39;Finished Training&#39;)</span>
</pre></div>
</div>
</div>
<div class="admonition note">
<p><strong>Note</strong> The above cell for <code class="docutils literal notranslate"><span class="pre">section</span> <span class="pre">2.2</span></code> needs to be run more than once by setting different parameters in the given <code class="docutils literal notranslate"><span class="pre">.json</span></code> file, if the correlation in <code class="docutils literal notranslate"><span class="pre">section</span> <span class="pre">3.4</span></code> is not satisfying. If one obtains a good performance, one can save the <code class="docutils literal notranslate"><span class="pre">dir_name</span></code> and then directly use it to load the trained model and infer gene expression, and doesn’t need to run <code class="docutils literal notranslate"><span class="pre">section</span> <span class="pre">2.2</span></code> again. Recommended parameter fine-tuning and its range: <code class="docutils literal notranslate"><span class="pre">training_epoch</span></code>: 40~60; <code class="docutils literal notranslate"><span class="pre">temperature</span></code>: 0.01-0.05.</p>
</div>
</section>
<section id="3.-Inference,-Imputation-and-Evaluation-on-within-spot">
<h2>3. Inference, Imputation and Evaluation on within spot<a class="headerlink" href="#3.-Inference,-Imputation-and-Evaluation-on-within-spot" title="Permalink to this heading"></a></h2>
<p>3.1 Infer the gene expression of within spots</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dir_name</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;/mnt/lingyu/nfs_share2/Python/FineST/FineST_local/Finetune/20241111220901614764&#39;
</pre></div></div>
</div>
<p>You can use the <code class="docutils literal notranslate"><span class="pre">dir_name</span></code> just obtained from the above cell, but here we use our trained <code class="docutils literal notranslate"><span class="pre">dir_name</span></code> for paper results repeated (saved in <code class="docutils literal notranslate"><span class="pre">FineST/FineST/Finetune/20240125140443830148</span></code>).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dir_name</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;FineST/FineST_local/Finetune/20240125140443830148&#39;</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># load params</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">parameter_file_path</span><span class="p">,</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">json_file</span><span class="p">:</span>
    <span class="n">params</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">json_file</span><span class="p">)</span>

<span class="c1"># load models</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">load_model</span><span class="p">(</span><span class="n">dir_name</span><span class="p">,</span> <span class="n">parameter_file_path</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">gene_hv</span><span class="p">)</span>
<span class="n">model</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>

<span class="c1"># load all data</span>
<span class="n">test_loader</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">build_loaders_inference</span><span class="p">(</span><span class="n">batch_size</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                          <span class="n">image_embed_path</span><span class="o">=</span><span class="n">image_embed_path_NPC</span><span class="p">,</span>
                                          <span class="n">spatial_pos_path</span><span class="o">=</span><span class="n">spatial_pos_path</span><span class="p">,</span>
                                          <span class="n">reduced_mtx_path</span><span class="o">=</span><span class="n">reduced_mtx_path</span><span class="p">,</span>
                                          <span class="n">dataset_class</span><span class="o">=</span><span class="s1">&#39;Visium&#39;</span><span class="p">)</span>


<span class="c1"># inference</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running inference tesk...&quot;</span><span class="p">)</span>

<span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

<span class="p">(</span><span class="n">matrix_profile</span><span class="p">,</span>
<span class="n">reconstructed_matrix</span><span class="p">,</span>
<span class="n">recon_ref_adata_image_f2</span><span class="p">,</span>
<span class="n">representation_image_reshape</span><span class="p">,</span>
<span class="n">representation_matrix</span><span class="p">,</span>
<span class="n">projection_image_reshape</span><span class="p">,</span>
<span class="n">projection_matrix</span><span class="p">,</span>
<span class="n">input_image_exp</span><span class="p">,</span>
<span class="n">reconstruction_iamge</span><span class="p">,</span>
<span class="n">reconstructed_matrix_reshaped</span><span class="p">,</span>
<span class="n">input_coord_all</span><span class="p">)</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">perform_inference_image</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">test_loader</span><span class="p">,</span> <span class="n">dataset_class</span><span class="o">=</span><span class="s1">&#39;Visium&#39;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;--- </span><span class="si">%s</span><span class="s2"> seconds for inference within spots ---&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start_time</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;reconstructed_matrix_reshaped shape: &quot;</span><span class="p">,</span> <span class="n">reconstructed_matrix_reshaped</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>

<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running inference tesk DONE!&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
***** Building loaders_inference *****
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
[2024-11-11 22:09:37] INFO - Running inference tesk...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Finished loading all files
***** Finished building loaders_inference *****
device cuda:0
***** Begin perform_inference: ******
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|█████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00&lt;00:00,  9.09it/s]
[2024-11-11 22:09:37] INFO - Running inference tesk DONE!
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
***** batch_size=adata.shape[0] doen&#39;t effect *****
torch.Size([1331, 596])
torch.Size([1331, 16, 384])
1
1
1
***** *****
Finished extractting test data
--- 0.1777350902557373 seconds for inference within spots ---
reconstructed_matrix_reshaped shape:  torch.Size([21296, 596])
</pre></div></div>
</div>
<p>3.2 Get coordinates of sub-spots from within spots</p>
<p>We use <code class="docutils literal notranslate"><span class="pre">reshape_latent_image()</span></code> function to convert the super-resolved gene expression with 16X relolution <code class="docutils literal notranslate"><span class="pre">reconstructed_matrix_reshaped</span></code> (torch.Size([21296, 596])) into tensor format <code class="docutils literal notranslate"><span class="pre">reconstructed_matrix_reshaped_tensor</span></code> (torch.Size([1331, 16, 596])).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">reconstructed_matrix_reshaped_tensor</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">reshape_latent_image</span><span class="p">(</span><span class="n">reconstructed_matrix_reshaped</span><span class="p">,</span>
                                                                   <span class="n">dataset_class</span><span class="o">=</span><span class="s1">&#39;Visium&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">reconstructed_matrix_reshaped_tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
torch.Size([1331, 16, 596])
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## plot the sub-spot level gene expression of the first gene in the first within spot</span>
<span class="p">(</span><span class="n">first_spot_first_variable</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span>
 <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">)</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">subspot_coord_expr_adata</span><span class="p">(</span><span class="n">reconstructed_matrix_reshaped_tensor</span><span class="p">,</span>
                                         <span class="n">adata</span><span class="p">,</span> <span class="n">gene_hv</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">q</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">dataset_class</span><span class="o">=</span><span class="s2">&quot;Visium&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;first_spot_first_variable shape:&quot;</span><span class="p">,</span> <span class="n">first_spot_first_variable</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">fstplt</span><span class="o">.</span><span class="n">subspot_expr</span><span class="p">(</span><span class="n">C</span><span class="p">,</span> <span class="n">first_spot_first_variable</span><span class="p">,</span> <span class="n">save_path</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
first_spot_first_variable shape: (16,)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/NPC_Train_Impute_53_1.png" src="_images/NPC_Train_Impute_53_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Get the sub-spot level gene expression of all genes in all within spots</span>
<span class="p">(</span><span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">all_spot_all_variable</span><span class="p">,</span>
 <span class="n">C2</span><span class="p">,</span> <span class="n">adata_spot</span><span class="p">)</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">subspot_coord_expr_adata</span><span class="p">(</span><span class="n">reconstructed_matrix_reshaped_tensor</span><span class="p">,</span>
                                                <span class="n">adata</span><span class="p">,</span> <span class="n">gene_hv</span><span class="p">,</span> <span class="n">dataset_class</span><span class="o">=</span><span class="s2">&quot;Visium&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;all_spot_all_variable shape:&quot;</span><span class="p">,</span> <span class="n">all_spot_all_variable</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata_spot</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
all_spot_all_variable shape: (21296, 596)
AnnData object with n_obs × n_vars = 21296 × 596
    obs: &#39;x&#39;, &#39;y&#39;
</pre></div></div>
</div>
<p>Here, <code class="docutils literal notranslate"><span class="pre">adata_spot</span></code> is the inferred super-resolved gene expression data with 16X solution.</p>
<p>3.3 Impute the gene expression of sub-spots</p>
<div class="line-block">
<div class="line">Input: two input files:</div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">adata</span></code>: original spot-level gene expression data (within spots).</div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">adata_spot</span></code>: inferred super-resolved gene expression data (within spots).</div>
</div>
<div class="line-block">
<div class="line">Input: three input files:</div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">sudo_adata</span></code>: Imputed data using the nearest <code class="docutils literal notranslate"><span class="pre">k</span></code> neighbors of within spots (here <code class="docutils literal notranslate"><span class="pre">k=6</span></code>).</div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">adata_impt</span></code>: <code class="docutils literal notranslate"><span class="pre">=</span> <span class="pre">w*adata_spot</span> <span class="pre">+</span> <span class="pre">(1-w)*sudo_adata</span></code>, predicted super-resolved gene expression data (here <code class="docutils literal notranslate"><span class="pre">w=0.5</span></code>).</div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">data_impt</span></code>: predicted super-resolved gene expression data in tensor form (within spots).</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sudo_adata</span><span class="p">,</span> <span class="n">adata_impt</span><span class="p">,</span> <span class="n">data_impt</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">impute_adata</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">adata_spot</span><span class="p">,</span> <span class="n">C2</span><span class="p">,</span> <span class="n">gene_hv</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">w</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sudo_adata</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata_impt</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
--- 4.314451217651367 seconds ---
AnnData object with n_obs × n_vars = 21296 × 596
    obs: &#39;x&#39;, &#39;y&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;mt&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;
AnnData object with n_obs × n_vars = 21296 × 596
    obs: &#39;x&#39;, &#39;y&#39;
</pre></div></div>
</div>
<p>For performance evaluation, we integrate super-resolved gene expression <code class="docutils literal notranslate"><span class="pre">data_impt</span></code> into spot-level gene expression <code class="docutils literal notranslate"><span class="pre">data_impt_reshape</span></code> (i.e., from 16x sub-spot level to spot level)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span><span class="p">,</span> <span class="n">data_impt_reshape</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">reshape_latent_image</span><span class="p">(</span><span class="n">data_impt</span><span class="p">,</span> <span class="n">dataset_class</span><span class="o">=</span><span class="s1">&#39;Visium&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;data_impt shape:&quot;</span><span class="p">,</span> <span class="n">data_impt</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;data_impt_reshape shape:&quot;</span><span class="p">,</span> <span class="n">data_impt_reshape</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
data_impt shape: torch.Size([21296, 596])
data_impt_reshape shape: torch.Size([1331, 596])
</pre></div></div>
</div>
<p>3.4 Visualization: selected gene (Original vs. FineST)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fstplt</span><span class="o">.</span><span class="n">gene_expr_compare</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;CD70&quot;</span><span class="p">,</span> <span class="n">data_impt_reshape</span><span class="p">,</span> <span class="n">gene_hv</span><span class="p">,</span> <span class="n">save_path</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/NPC_Train_Impute_62_0.png" src="_images/NPC_Train_Impute_62_0.png" />
</div>
</div>
<p>3.4 Correlation: Selected gene、all spots and all genes、mean correlation box</p>
<div class="line-block">
<div class="line">The correlation can be used as a measure to evaluate the gene expression prediction performance. Here we provide three functions:</div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">sele_gene_cor</span></code>: show the selected gene correlation (Reconstructed by FineST vs. Original).</div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">mean_cor</span></code> : calculate the mean correlation of all spots or all genes.</div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">mean_cor_box</span></code> : box plot of correlation for all spots or all genes.</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## if ones wants to save the figure in .pdf format, just set: save_path=str(gene)+&#39;correlation.pdf&#39;</span>
<span class="n">fstplt</span><span class="o">.</span><span class="n">sele_gene_cor</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">data_impt_reshape</span><span class="p">,</span> <span class="n">gene_hv</span><span class="p">,</span> <span class="n">gene</span> <span class="o">=</span> <span class="s2">&quot;CD70&quot;</span><span class="p">,</span>
                     <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;FineST Expression&#39;</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;CD70 expression&quot;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
                     <span class="n">save_path</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/NPC_Train_Impute_65_0.png" src="_images/NPC_Train_Impute_65_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running correlation task...&quot;</span><span class="p">)</span>

<span class="p">(</span><span class="n">pearson_cor_spot</span><span class="p">,</span>
 <span class="n">spearman_cor_spot</span><span class="p">,</span>
 <span class="n">cosine_sim_spot</span><span class="p">)</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">mean_cor</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">data_impt_reshape</span><span class="p">,</span> <span class="s1">&#39;reconf2&#39;</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="s2">&quot;spot&quot;</span><span class="p">)</span>

<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running correlation task DINE!&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
[2024-11-11 22:09:43] INFO - Running correlation task...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
matrix1:  (1331, 596)
matrix2:  (1331, 596)
Mean Pearson correlation coefficient--reconf2: 0.8535
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
[2024-11-11 22:09:44] INFO - Running correlation task DINE!
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Mean Spearman correlation coefficient--reconf2: 0.6093
Mean cosine similarity--reconf2: 0.6075
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running Gene Correlation task...&quot;</span><span class="p">)</span>

<span class="p">(</span><span class="n">pearson_cor_gene</span><span class="p">,</span>
 <span class="n">spearman_cor_gene</span><span class="p">,</span>
 <span class="n">cosine_sim_gene</span><span class="p">)</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">mean_cor</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">data_impt_reshape</span><span class="p">,</span> <span class="s1">&#39;reconf2&#39;</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>

<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running Gene Correlation task DINE!&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
[2024-11-11 22:09:44] INFO - Running Gene Correlation task...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
matrix1:  (1331, 596)
matrix2:  (1331, 596)
Mean Pearson correlation coefficient--reconf2: 0.8845
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
[2024-11-11 22:09:44] INFO - Running Gene Correlation task DINE!
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Mean Spearman correlation coefficient--reconf2: 0.5380
Mean cosine similarity--reconf2: 0.6075
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fstplt</span><span class="o">.</span><span class="n">mean_cor_box</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">data_impt_reshape</span><span class="p">,</span> <span class="n">save_path</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
0.8534651812923978
0.8845136777311444
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/NPC_Train_Impute_68_1.png" src="_images/NPC_Train_Impute_68_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## save adata</span>
<span class="c1"># patientxy = &#39;patient1&#39;</span>
<span class="c1"># adata_impt.write_h5ad(str(path)+&#39;FineST/FineST_local/Dataset/ImputData/&#39;+str(patientxy)+&#39;/&#39;+str(patientxy)+&#39;_adata.h5ad&#39;)</span>
</pre></div>
</div>
</div>
<div class="admonition note">
<p><strong>Note</strong> Here the <code class="docutils literal notranslate"><span class="pre">adata_impt</span></code> is the imputed spot-level gene expression from FineST, it only contains the given 1331 within spots. The comparisons are based on the original adata <code class="docutils literal notranslate"><span class="pre">patient1_adata_orignal.h5ad</span></code> vs. FineST’s predicted adata <code class="docutils literal notranslate"><span class="pre">patient1_adata.h5ad</span></code>.</p>
</div>
</section>
<section id="4.-Infer-gene-expression-of-%22within-spot%22-and-%22between-spot%22">
<h2>4. Infer gene expression of “within spot” and “between spot”<a class="headerlink" href="#4.-Infer-gene-expression-of-%22within-spot%22-and-%22between-spot%22" title="Permalink to this heading"></a></h2>
<p>4.1 Get all spot coordinates (within spot &amp; between spot) of image embeddings</p>
<div class="line-block">
<div class="line">Input: two input file pathways:</div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">file_paths_spot</span></code>: image embedding, of within spots, from <strong>Setp0</strong> <code class="docutils literal notranslate"><span class="pre">HIPT_image_feature_extract.py</span></code>.</div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">file_paths_between_spot</span></code>: image embedding, of between spots, from <strong>Setp0</strong> <code class="docutils literal notranslate"><span class="pre">HIPT_image_feature_extract.py</span></code>.</div>
</div>
<p>For NPC dataset, the image embedding file can be downloaded <a class="reference external" href="https://drive.google.com/drive/folders/1ypaTohLM2lP4EnKeYTS6QUvv-ymcRVZo">Goole Drive</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## add coords for each .pth file</span>
<span class="n">file_paths_spot</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;NPC/Data/stdata/ZhuoLiang/LLYtest/AH_Patient1_pth_64_16/&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;file_paths_spot number: &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">file_paths_spot</span><span class="p">))</span>
<span class="n">file_paths_between_spot</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;NPC/Data/stdata/ZhuoLiang/LLYtest/NEW_AH_Patient1_pth_64_16/&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;file_paths_between_spot number:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">file_paths_between_spot</span><span class="p">))</span>
<span class="n">file_paths_all</span> <span class="o">=</span> <span class="n">file_paths_spot</span> <span class="o">+</span> <span class="n">file_paths_between_spot</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;file_paths_all number:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">file_paths_all</span><span class="p">))</span>

<span class="c1">## Merge, sort and process file paths</span>
<span class="n">data_all</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">get_image_coord_all</span><span class="p">(</span><span class="n">file_paths_all</span><span class="p">,</span> <span class="n">dataset_class</span><span class="o">=</span><span class="s1">&#39;Visium&#39;</span><span class="p">)</span>
<span class="n">position_order_allspot</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data_all</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;pixel_y&#39;</span><span class="p">,</span> <span class="s1">&#39;pixel_x&#39;</span><span class="p">])</span>
<span class="n">position_order_allspot</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
file_paths_spot number:  1331
file_paths_between_spot number: 3708
file_paths_all number: 5039
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pixel_y</th>
      <th>pixel_x</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>10014</td>
      <td>10023</td>
    </tr>
    <tr>
      <th>1</th>
      <td>10015</td>
      <td>9649</td>
    </tr>
    <tr>
      <th>2</th>
      <td>10016</td>
      <td>9276</td>
    </tr>
    <tr>
      <th>3</th>
      <td>10020</td>
      <td>7783</td>
    </tr>
    <tr>
      <th>4</th>
      <td>10021</td>
      <td>7410</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>5034</th>
      <td>9978.0</td>
      <td>3023.5</td>
    </tr>
    <tr>
      <th>5035</th>
      <td>9978.5</td>
      <td>2836.5</td>
    </tr>
    <tr>
      <th>5036</th>
      <td>9979.0</td>
      <td>2650.0</td>
    </tr>
    <tr>
      <th>5037</th>
      <td>9980.0</td>
      <td>2277.0</td>
    </tr>
    <tr>
      <th>5038</th>
      <td>9980.5</td>
      <td>2090.5</td>
    </tr>
  </tbody>
</table>
<p>5039 rows × 2 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## save all spots</span>
<span class="c1"># position_order_allspot.to_csv(str(path)+&quot;FineST/FineST/Dataset/NPC/ContrastP1geneLR/position_order_all.csv&quot;, index=False, header=False)</span>
</pre></div>
</div>
</div>
<div class="admonition note">
<p><strong>Note</strong> The above two cells for <code class="docutils literal notranslate"><span class="pre">section</span> <span class="pre">4.1</span></code> only need to be run once, aiming to generate the ordered ST spot coordinates <code class="docutils literal notranslate"><span class="pre">position_order_all.csv</span></code> according to image pixel coordinates.</p>
</div>
<p>4.2 Load all spot image embeddings with their coordinates</p>
<div class="line-block">
<div class="line">Input: three input file pathway:</div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">file_paths_spot</span></code>: image embedding, of within spots, from <strong>Setp0</strong> <code class="docutils literal notranslate"><span class="pre">HIPT_image_feature_extract.py</span></code>.</div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">file_paths_between_spot</span></code>: image embedding, of between spots, from <strong>Setp0</strong> <code class="docutils literal notranslate"><span class="pre">HIPT_image_feature_extract.py</span></code>.</div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">spatial_pos_path</span></code>: ordered all ST spot coordinates from above cell, can be found in <code class="docutils literal notranslate"><span class="pre">FineST/Dataset/NPC/ContrastP1geneLR/position_order_all.csv</span></code></div>
</div>
<p>For NPC dataset, the image embedding file can be downloaded <a class="reference external" href="https://drive.google.com/drive/folders/1ypaTohLM2lP4EnKeYTS6QUvv-ymcRVZo">Goole Drive</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">glob</span>
<span class="n">file_paths_spot</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;NPC/Data/stdata/ZhuoLiang/LLYtest/AH_Patient1_pth_64_16/*.pth&#39;</span>
<span class="n">file_paths_between_spot</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;NPC/Data/stdata/ZhuoLiang/LLYtest/NEW_AH_Patient1_pth_64_16/*.pth&#39;</span>
<span class="n">spatial_pos_path</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;FineST/FineST_local/Dataset/NPC/ContrastP1geneLR/position_order_all.csv&#39;</span>
</pre></div>
</div>
</div>
<p>4.4 Load the trained model to infer all spots</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## load params</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">parameter_file_path</span><span class="p">,</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">json_file</span><span class="p">:</span>
    <span class="n">params</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">json_file</span><span class="p">)</span>

<span class="c1">## load models</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">load_model</span><span class="p">(</span><span class="n">dir_name</span><span class="p">,</span> <span class="n">parameter_file_path</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">gene_hv</span><span class="p">)</span>
<span class="n">model</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>
<span class="c1"># print(&quot;model&quot;, model)</span>

<span class="c1">## load all spot (within and between) spots data</span>
<span class="n">all_dataset</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">build_loaders_inference_allimage</span><span class="p">(</span><span class="n">batch_size</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">file_paths_all</span><span class="p">),</span>
                                                   <span class="n">file_paths_spot</span><span class="o">=</span><span class="n">image_embed_path_NPC</span><span class="p">,</span>
                                                   <span class="n">file_paths_between_spot</span><span class="o">=</span><span class="n">file_paths_between_spot</span><span class="p">,</span>
                                                   <span class="n">spatial_pos_path</span><span class="o">=</span><span class="n">spatial_pos_path</span><span class="p">,</span>
                                                   <span class="n">dataset_class</span><span class="o">=</span><span class="s1">&#39;Visium&#39;</span><span class="p">)</span>

<span class="c1">## inference</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running inference tesk between spot...&quot;</span><span class="p">)</span>

<span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

<span class="p">(</span><span class="n">recon_ref_adata_image_f2</span><span class="p">,</span>
<span class="n">reconstructed_matrix_reshaped</span><span class="p">,</span>
<span class="n">representation_image_reshape_between_spot</span><span class="p">,</span>
<span class="n">input_image_exp_between_spot</span><span class="p">,</span>
<span class="n">input_coord_all</span><span class="p">)</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">perform_inference_image_between_spot</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">all_dataset</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;--- </span><span class="si">%s</span><span class="s2"> seconds ---&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start_time</span><span class="p">))</span>

<span class="c1">## print</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;recon_ref_adata_image_f2:&quot;</span><span class="p">,</span> <span class="n">recon_ref_adata_image_f2</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="c1"># print(&quot;recon_ref_adata_image_f2:\n&quot;, recon_ref_adata_image_f2)</span>
<span class="c1"># print(&quot;input_coord_all:\n&quot;, input_coord_all)</span>

<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running inference tesk between spot DONE!&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
***** Building loaders_inference between spot *****
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
[2024-11-11 22:09:49] INFO - Running inference tesk between spot...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Finished loading all files
***** Finished building loaders_inference between spot *****
device cuda:0
***** Begin perform_inference: ******
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|█████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00&lt;00:00,  3.71it/s]
[2024-11-11 22:09:49] INFO - Running inference tesk between spot DONE!
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
***** batch_size=adata.shape[0] *****
torch.Size([5039, 16, 384])
1
***** *****
Finished extractting image_between_spot data
--- 0.3137013912200928 seconds ---
recon_ref_adata_image_f2: (5039, 596)
</pre></div></div>
</div>
<p>4.5 Visualization all spots</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## process_and_check_duplicates 0f the coordinates of all spots (within and between)</span>
<span class="n">spatial_loc_all</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">get_allspot_coors</span><span class="p">(</span><span class="n">input_coord_all</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">spatial_loc_all</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Are there any duplicate rows? : False
[[10014.  10023. ]
 [10015.   9649. ]
 [10016.   9276. ]
 ...
 [ 9979.   2650. ]
 [ 9980.   2277. ]
 [ 9980.5  2090.5]]
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># fstplt.gene_expr_allspots(&quot;CD70&quot;, spatial_loc_all, recon_ref_adata_image_f2, gene_hv, &#39;Inferred all spot&#39;, s=4, save_path=None)</span>
</pre></div>
</div>
</div>
<p>4.6 Visualization all sub-spots</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">reconstructed_matrix_reshaped_tensor</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">reshape_latent_image</span><span class="p">(</span><span class="n">reconstructed_matrix_reshaped</span><span class="p">,</span>
                                                                   <span class="n">dataset_class</span><span class="o">=</span><span class="s1">&#39;Visium&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">reconstructed_matrix_reshaped_tensor</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
torch.Size([5039, 16, 596])
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Get the sub-spot level gene expression of all genes in all spots</span>
<span class="p">(</span><span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">all_spot_all_variable</span><span class="p">,</span>
 <span class="n">C2</span><span class="p">,</span> <span class="n">adata_spot_all</span><span class="p">)</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">subspot_coord_expr_adata</span><span class="p">(</span><span class="n">reconstructed_matrix_reshaped_tensor</span><span class="p">,</span>
                                                    <span class="n">spatial_loc_all</span><span class="p">,</span> <span class="n">gene_hv</span><span class="p">,</span> <span class="n">dataset_class</span><span class="o">=</span><span class="s2">&quot;Visium&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;all_spot_all_variable shape:&quot;</span><span class="p">,</span> <span class="n">all_spot_all_variable</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata_spot_all</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
all_spot_all_variable shape: (80624, 596)
AnnData object with n_obs × n_vars = 80624 × 596
    obs: &#39;x&#39;, &#39;y&#39;
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># fstplt.gene_expr_allspots(&quot;CD70&quot;, C2, all_spot_all_variable, gene_hv, &#39;Inferred all spot&#39;, s=0.5, save_path=None)</span>
</pre></div>
</div>
</div>
</section>
<section id="5.-Imputate-sub-spot-gene-expression-using-measured-spot-expression">
<h2>5. Imputate sub-spot gene expression using measured spot expression<a class="headerlink" href="#5.-Imputate-sub-spot-gene-expression-using-measured-spot-expression" title="Permalink to this heading"></a></h2>
<p>5.1 Impute the gene expression of all sub-spots from all spots (within spots and between spots)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sudo_adata_all</span><span class="p">,</span> <span class="n">adata_impt_all</span><span class="p">,</span> <span class="n">data_impt_all</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">impute_adata</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">adata_spot_all</span><span class="p">,</span> <span class="n">C2</span><span class="p">,</span>
                                                                 <span class="n">gene_hv</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">w</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sudo_adata_all</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata_impt_all</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
--- 16.544437885284424 seconds ---
AnnData object with n_obs × n_vars = 80624 × 596
    obs: &#39;x&#39;, &#39;y&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;mt&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;
AnnData object with n_obs × n_vars = 80624 × 596
    obs: &#39;x&#39;, &#39;y&#39;
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## save adata: 80624 × 596</span>
<span class="c1"># patientxy = &#39;patient1&#39;</span>
<span class="c1"># adata_impt_all.write_h5ad(str(path)+&#39;FineST/FineST_local/Dataset/ImputData/&#39;+str(patientxy)+&#39;/&#39;+str(patientxy)+&#39;_adata_all.h5ad&#39;)</span>
</pre></div>
</div>
</div>
<div class="admonition note">
<p><strong>Note</strong> Here, <code class="docutils literal notranslate"><span class="pre">adata_impt_all</span></code> is the imputed sub-spot-level gene expression from FineST. <code class="docutils literal notranslate"><span class="pre">patient1_adata_all.h5ad</span></code> contains about 1331x4x16 sub-spots. (1331 within spots, 1331x3 between spots, 16x resolution).</p>
</div>
<p>5.2 Convert super-resolved gene expression to spot-level and save the imputated spot-level adata</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span><span class="p">,</span> <span class="n">adata_impt_all_reshape</span> <span class="o">=</span> <span class="n">fst</span><span class="o">.</span><span class="n">reshape_latent_image</span><span class="p">(</span><span class="n">data_impt_all</span><span class="p">,</span> <span class="n">dataset_class</span><span class="o">=</span><span class="s1">&#39;Visium&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;data_impt_all shape:&quot;</span><span class="p">,</span> <span class="n">adata_impt_all</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;adata_impt_all_reshape shape:&quot;</span><span class="p">,</span> <span class="n">adata_impt_all_reshape</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
data_impt_all shape: (80624, 596)
adata_impt_all_reshape shape: torch.Size([5039, 596])
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_impt_spot</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">X</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata_impt_all_reshape</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()))</span>
<span class="n">adata_impt_spot</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">gene_hv</span>
<span class="n">adata_impt_spot</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">spatial_loc_all</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span>
<span class="n">adata_impt_spot</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">spatial_loc_all</span><span class="p">[:,</span><span class="mi">1</span><span class="p">]</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;adata_impt_spot: &quot;</span><span class="p">,</span> <span class="n">adata_impt_spot</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata_impt_spot</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
adata_impt_spot:  AnnData object with n_obs × n_vars = 5039 × 596
    obs: &#39;x&#39;, &#39;y&#39;
0       10014.0
1       10015.0
2       10016.0
3       10020.0
4       10021.0
         ...
5034     9978.0
5035     9978.5
5036     9979.0
5037     9980.0
5038     9980.5
Name: x, Length: 5039, dtype: float64
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[51]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## save data: 5039 × 596</span>
<span class="c1"># patientxy = &#39;patient1&#39;</span>
<span class="c1"># adata_impt_spot.write_h5ad(str(path)+&#39;FineST/FineST_local/Dataset/ImputData/&#39;+str(patientxy)+&#39;/&#39;+str(patientxy)+&#39;_adata_all_spot.h5ad&#39;)</span>
</pre></div>
</div>
</div>
<div class="admonition note">
<p><strong>Note</strong> Here, <code class="docutils literal notranslate"><span class="pre">adata_impt_spot</span></code> is the imputed sub-spot-level gene expression from FineST. <code class="docutils literal notranslate"><span class="pre">patient1_adata_all_spot.h5ad</span></code> contains about 1331x4 spots. (1331 within spots, 1331x3 between spots), it is integrated from <code class="docutils literal notranslate"><span class="pre">patient1_adata_all.h5ad</span></code> using the <code class="docutils literal notranslate"><span class="pre">reshape_latent_image</span></code> function.</p>
</div>
<p>5.3 Visualization: gene at all spot</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fstplt</span><span class="o">.</span><span class="n">gene_expr_allspots</span><span class="p">(</span><span class="s2">&quot;CD70&quot;</span><span class="p">,</span> <span class="n">spatial_loc_all</span><span class="p">,</span> <span class="n">adata_impt_all_reshape</span><span class="p">,</span> <span class="n">gene_hv</span><span class="p">,</span>
                          <span class="s1">&#39;FineST all spot&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">save_path</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
CD70 gene expression dim:  (5039, 1)
CD70 gene expression:
 [[0.31857272]
 [0.38838257]
 [0.3760277 ]
 ...
 [0.49542667]
 [0.55092828]
 [0.54530712]]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/NPC_Train_Impute_100_1.png" src="_images/NPC_Train_Impute_100_1.png" />
</div>
</div>
<p>5.4 Visualization: gene at all sub-spot</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[53]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fstplt</span><span class="o">.</span><span class="n">gene_expr_allspots</span><span class="p">(</span><span class="s2">&quot;CD70&quot;</span><span class="p">,</span> <span class="n">C2</span><span class="p">,</span> <span class="n">adata_impt_all</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">gene_hv</span><span class="p">,</span>
                          <span class="s1">&#39;FineST all sub-spot&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">save_path</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
CD70 gene expression dim:  (80624, 1)
CD70 gene expression:
 [[0.2771624 ]
 [0.26723772]
 [0.25254893]
 ...
 [0.60330289]
 [0.59376435]
 [0.58118813]]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/NPC_Train_Impute_102_1.png" src="_images/NPC_Train_Impute_102_1.png" />
</div>
</div>
<p>For NPC dataset, the the above four <code class="docutils literal notranslate"><span class="pre">.h5ad</span></code> adata files can be downloaded from <a class="reference external" href="https://drive.google.com/drive/folders/1ypaTohLM2lP4EnKeYTS6QUvv-ymcRVZo">Goole Drive</a>.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Crop_ROI_image.html" class="btn btn-neutral float-left" title="Crop image of ROI" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="CRC16_Train_Impute.html" class="btn btn-neutral float-right" title="FineST in CRC Visium HD data" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Lingyu Li.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>